src/messenger/messenger.cpp

```diff
@@ -0,0 +1,256 @@
+ *    This is ZeroMQ terminology for saying that:
+ *    1.  The server process exposes one main ROUTER socket.
+ *    2.  The ROUTER socket is a "server" socket that asynchronously sends and receives messages to "clients".
+ *        To help the ROUTER route messages, every message is expected to be multipart and of the following form:
+ *        ROUTING_IDENTIFIER DELIMITER PAYLOAD
+ *        where:
+ *          - ROUTING_IDENTIFIER is controlled by setsockopt or getsockopt on ZMQ_ROUTING_ID.
+ *          - DELIMITER is an empty message with size 0.
+ *          - PAYLOAD is the message itself.
+ *    3.  The DEALER socket must send messages in the same format as well.
+ *
+ * # Interfacing with the Messenger from Python.
+ *
+ *    This is an example of how you connect to the Messenger from Python.
+ *      import zmq                                    # Use the ZeroMQ library.
+ *      ctx = zmq.Context()                           # Create a ZeroMQ context for our entire process.
+ *      sock = ctx.socket(zmq.DEALER)                 # We want an async DEALER socket for reasons described above.
+ *      sock.setsockopt(zmq.IDENTITY, b'snek')        # Set the name of our Python program,
+ *      sock.connect('ipc:///tmp/noisepage-ipc0')     # Connect to NoisePage on the same machine over IPC.
+ *      sock.send(b'', flags=zmq.SNDMORE)             # Start building a message. This is the empty delimiter packet.
+ *      s.send(b'PThis is the message payload.')      # Finish the message and send it. P is a prefix for Print.
+ *
+ */
+
+namespace terrier::messenger {
+
+/** An abstraction around ZeroMQ messages which explicitly have the sender specified. */
+class ZmqMessage {
+ public:
+  /** The routing ID of the message sender. */
+  std::string identity_;
+  /** The payload in the message. */
+  std::string payload_;
+};
+
+}  // namespace terrier::messenger
+
+namespace terrier {
+
+/**
+ * Useful ZeroMQ utility functions implemented in a naive manner. Most functions have wasteful copies.
+ * If perf indicates that these functions are a bottleneck, switch to the zero-copy messages of ZeroMQ.
+ */
+class ZmqUtil {
+ public:
+  /** ZmqUtil is a static utility class that should not be instantiated. */
+  ZmqUtil() = delete;
+
+  /** The maximum length of a routing ID. Specified by ZeroMQ. */
+  static constexpr int MAX_ROUTING_ID_LEN = 255;
+
+  /** @return The routing ID of the socket. */
+  static std::string GetRoutingId(common::ManagedPointer<zmq::socket_t> socket) {
+    char buf[MAX_ROUTING_ID_LEN];
+    size_t routing_id_len = MAX_ROUTING_ID_LEN;
+    socket->getsockopt(ZMQ_ROUTING_ID, &buf, &routing_id_len);
+    return std::string(buf, routing_id_len);
+  }
+
+  /** @return The next string to be read off the socket. */
+  static std::string Recv(common::ManagedPointer<zmq::socket_t> socket, zmq::recv_flags flags) {
+    zmq::message_t message;
+    auto received = socket->recv(message, flags);
```

Comment:
This call may raise errors when the serverloop terminates (because I believe this `recv` is a blocking call), and it will raise errors 'host not reachable'. 
What is the best way to handler someone calls `messenger->Terminate()` when it is blocked in this `recv` call?

Reply:
Is `messenger->Terminate()` able to be called? If so, I can add an explicit call to `zmq_ctx->shutdown()` which causes all blocking socket operations to end with `ETERM`.

Reply:
Resolved in latest push.